shark_turbine/kernel/functional/README.md

```diff
@@ -0,0 +1,99 @@
+
+In GPU programming, a wavefront is a group of threads (or work items) that execute the same instruction in lockstep. In particular, coalesced memory accesses by all threads in a wavefront are executed together. This is analogous to the concept of a "warp" in NVIDIA's CUDA programming model.
+Typically, on AMD GPUs, a wavefront contains 32 or 64 threads, all of which participate in executing the same instruction at the same time.
+
+In this representation, memory access patterns are more naturally optimized for coalescing, reducing the complexity and manual effort required to achieve optimized memory transactions. In consequence, programmers can focus more on the core computational logic rather than the intricacies of thread coordination and memory management.
+This approach contrasts with traditional models like OpenCL and CUDA, which often require more explicit management of threads and synchronization.
+
+## Gemm example
+
+```Python
+# Wave tile sizes
+M = tkl.sym.M
+N = tkl.sym.N
+K = tkl.sym.K
+# Workgroup tile sizes
+BLOCK_M = tkl.sym.BLOCK_M
+BLOCK_N = tkl.sym.BLOCK_N
+BLOCK_K = tkl.sym.BLOCK_K
+# Address space (for GPU, shared(1) or global(0))
+ADDRESS_SPACE = tkl.sym.ADDRESS_SPACE
+# Other hyperparameters
+LOAD_ELEMS_PER_THREAD = tkl.sym.LOAD_ELEMS_PER_THREAD
+STORE_ELEMS_PER_THREAD = tkl.sym.STORE_ELEMS_PER_THREAD
+
+# Expose user-constraints
+constraints = [tkf.WorkgroupConstraint(M, BLOCK_M, 0)]
+constraints += [tkf.WorkgroupConstraint(N, BLOCK_N, 1)]
+constraints += [tkf.TilingConstraint(K, BLOCK_K)]
+constraints += [tkf.WaveConstraint(M, BLOCK_M, 0, 64)]
+constraints += [tkf.WaveConstraint(N, BLOCK_N, 1, 64)]
+constraints += [
+    tkf.HardwareConstraint(
+        threads_per_wave=64, mma_type="MFMA_F32_16x16x16_F16"
+    )
+]
+
+# Wave-level micro-kernel.
+# Since warps are not directly addressable, there is no
+# explicit notion of a warp id (like a workgroup or thread id).
+@tkf.wave(constraints)
+def gemm(
+    a: tkl.Memory[M, K, ADDRESS_SPACE, tkl.f16],
+    b: tkl.Memory[N, K, ADDRESS_SPACE, tkl.f16],
+    c: tkl.Memory[M, N, ADDRESS_SPACE, tkl.f32],
+):
+    # c_reg: tkf.Register[M, N, tkl.f32]
+    c_reg = tkf.construct_register_from_metadata((M, N), tkl.f32, 0.0)
+
+    # This microkernel encodes the fact that if the reduction
+    # dimension were tiled, then we would need to materialize a loop.
+    @tkf.tiled_loop(K, init_args=[c_reg])
+    def repeat(c_reg) -> tkl.Register[M, N, tkl.f32]:
+        # a_reg: tkf.Register[M, K, tkl.f16]
+        a_reg = tkf.read(a, elements_per_thread=LOAD_ELEMS_PER_THREAD)
+        # b_reg: tkf.Register[N, K, tkl.f16]
+        b_reg = tkf.read(b, elements_per_thread=LOAD_ELEMS_PER_THREAD)
+        # c_reg: tkf.Register[M, N, tkl.f32]
+        c_reg = tkf.mma(a_reg, b_reg, c_reg)
+        return c_reg
+
+    # repeat represents the results of the loop
+    tkf.write(repeat, c, elements_per_thread=STORE_ELEMS_PER_THREAD)
```

Comment:
A minor syntax issue: here it seems like we have just defined a repeat function and are passing that function to tkf.write. Instead we might want to call repeat and get a result and then pass that to tkf.write. Something like
```
result = repeat(c_reg)
tkf.write(result, c, ...)
```
What do you think?